.github/actions/generate-attestations/src/main.ts

```diff
@@ -0,0 +1,58 @@
+import * as core from "@actions/core";
+import fs from "fs";
+import path from "path";
+import { writeAttestations } from "./attestation";
+
+// Detect directory traversal for input file.
+function resolvePathInput(input: string, wd: string): string {
+  const safeJoin = path.join(wd, input);
+  if (!safeJoin.startsWith(wd)) {
+    throw Error(`unsafe path ${safeJoin}`);
+  }
+  return safeJoin;
+}
+
+export async function run(): Promise<void> {
+  try {
+    const wd = process.env[`GITHUB_WORKSPACE`] || "";
+
+    // SLSA subjects layout file.
+    const slsaOutputs = core.getInput("slsa-outputs-file");
+    const safeSlsaOutputs = resolvePathInput(slsaOutputs, wd);
+    core.debug(`Using SLSA output file at ${safeSlsaOutputs}!`);
+
+    // Predicate.
+    const predicateFile = core.getInput("predicate-file");
+    const safePredicateFile = resolvePathInput(predicateFile, wd);
+    core.debug(`Inputs: Predicate file ${safePredicateFile}!`);
+
+    // Predicate type
+    const predicateType = core.getInput("predicate-type");
+    core.debug(`Inputs: Predicate type ${predicateType}!`);
+
+    // Attach subjects and generate attestation files
+    const outputFolder = core.getInput("output-folder");
+    const attestations = await writeAttestations(
+      safeSlsaOutputs,
+      predicateType,
+      safePredicateFile
+    );
+
+    // Write attestations
+    fs.mkdirSync(outputFolder, { recursive: true });
+    for (const att in attestations) {
+      const outputFile = `${outputFolder}/${att}`;
```

Comment:
use `path.join()` and validate for path traversal?

Reply:
done! factored the path validation out into a function

Reply:
fyi, this is the similar Go function we use in case you want to compare https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/utils/path_test.go#L27